packages/core/core-v2/src/Parcel.js

```diff
@@ -0,0 +1,189 @@
+// @flow
+'use strict';
+import path from 'path';
+import { AbortController } from 'abortcontroller-polyfill/dist/cjs-ponyfill';
+import Watcher from '@parcel/watcher';
+import PQueue from 'p-queue';
+import AssetGraph from './AssetGraph';
+import { Node } from './Graph';
+import type { Dependency, Asset, File } from './types';
+import TransformerRunner from './TransformerRunner';
+import ResolverRunner from './ResolverRunner';
+import BundlerRunner from './BundlerRunner';
+import PackagerRunner from './PackagerRunner';
+
+// TODO: use custom config if present
+const defaultConfig = require('@parcel/config-default');
+
+const AbortError = new Error('Aborted!');
+
+type CliOpts = {
+  watch?: boolean
+}
+
+type ParcelOpts = {
+  entries: Array<string>,
+  cliOpts: CliOpts,
+}
+
+type Signal = {
+  aborted: boolean,
+  addEventListener?: Function,
+}
+
+type BuildOpts = {
+  signal: Signal,
+  shallow?: boolean,
+}
+
+export default class Parcel {
+  rootDir: string;
+  graph: AssetGraph;
+  watcher: Watcher;
+  updateQueue: PQueue;
+  mainQueue: PQueue;
+  transformerRunner: TransformerRunner;
+  resolverRunner: ResolverRunner;
+  bundlerRunner: BundlerRunner;
+  packagerRunner: PackagerRunner;
+
+  constructor({
+    entries,
+    cliOpts = {},
+  }: ParcelOpts) {
+    this.rootDir = process.cwd();
+
+    this.graph = new AssetGraph({ entries, rootDir: this.rootDir });
+    this.watcher = cliOpts.watch ? new Watcher() : null;
+    this.updateQueue = new PQueue({ autoStart: false });
+    this.mainQueue = new PQueue({ autoStart: false });
+
+    this.transformerRunner = new TransformerRunner({ parcelConfig: defaultConfig, cliOpts });
+    this.resolverRunner = new ResolverRunner();
+    this.bundlerRunner = new BundlerRunner({ parcelConfig: defaultConfig, cliOpts });
+    this.packagerRunner = new PackagerRunner({ parcelConfig: defaultConfig, cliOpts });
+  }
+
+  run() {
+    let controller = new AbortController();
+    let signal = controller.signal;
+
+    this.build({ signal });
```

Comment:
Is there a reason for not awaiting the promise this returns later on?

Reply:
I was probably thinking that it made more sense to start watching after a build started and using an await would mean it wouldn't start watching until the first build is finished. It doesn't really need to wait on the build to do anything after, but I suppose it might in the future. Is that a compelling enough reason to switch things around and use an await even if it's not doing anything after yet?

Reply:
I guess it's more about the run function returning a promise that only resolves when it's not running anymore. I'll work on that.

Reply:
You could write:

```js
let promise = this.build();
// ...
await promise;
```